### PR TITLE
Fixed randomization problem with masking 

### LINKED ISSUE #492: Loss oscilation during ERA5 training
### What happened?

Training loss is oscilating with high frequency as shown on the plot below

Image

### What are the steps to reproduce the bug?

Train for couple of epochs 

### Version

0.0.1

### Platform (OS and architecture)

Linux

### Relevant log output

```shell

```

### Accompanying data

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by clessig**
## Description

Fixed randomization problem with masking

## Type of Change

-   [X] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #492

## Code Compatibility

-   [X] I have performed a self-review of my code

### Code Performance and Testing

-   [X] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by clessig** [comment]
This is with 2.5 epochs for 2 nodes = 8 DDP workers:

s03mn46x_k4hiy385_q8ow3kvj_i14tiowq_t039q2te_wkbn2v9t_train_val_ERA5

Also no repetitive pattern, so as far as I can tell the problem is fixed.

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 1]
```diff
@@ -413,6 +420,19 @@ def worker_workset(self):
...
         else:
+            # ensure the rng seed is fully unique across workers and epochs
+            # the worker processes are generated as bit-wise copy of the "template" (the actual
+            # instance of the present class that is created) whenever __iter__ is started. This
+            # happens for each epoch, for train and validation, and independently for each DDP
+            # worker. After the bit-wise copy, the rng seed needs to be made unique for
+            # DDP workers, loader process, epoch.
+            dist = torch.distributed
+            self.data_loader_rng_seed *= (
+                (((dist.get_rank() + 1) * 73) if dist.is_initialized() else 1)
```
Comment (sophie-xhonneux) [inline src/weathergen/datasets/multi_stream_data_sampler.py:432]: is there a particular logic why we picked these numbers? they aren't all prime, which would make more sense to me

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 2]
```diff
@@ -173,18 +173,24 @@ def __init__(
...
 
         self.batch_size = batch_size
-        self.rng = np.random.default_rng(cf.data_loader_rng_seed)
+
+        self.data_loader_rng_seed = (
+            cf.data_loader_rng_seed
+            if cf.data_loader_rng_seed > cf.loader_num_workers
+            else cf.data_loader_rng_seed * 13
+        )
+        self.rng = np.random.default_rng(self.data_loader_rng_seed)
```
Comment (sophie-xhonneux) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: Sorry if this is a stupid question, but shouldn't the seed also depend on the epoch? the way I understand the code right now we reset the self.rng to the same np.random with the same seed every epoch no?

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 1 above)
Comment (clessig) [reply src/weathergen/datasets/multi_stream_data_sampler.py:432]: Yes, should all be primes. Fixed.

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 2 above)
Comment (clessig) [reply src/weathergen/datasets/multi_stream_data_sampler.py]: The self.rng initalization here has been removed now, since the rng for and epoch is always created in `reset()`.

The epoch dependence is introduced in `worker_workset()`. The seed here is only used when one runs with `num_loader_workers = 0`, which is only for debug purposes and the randomization is not relevant.
